File: apps/desktop/src-tauri/src/commands/sync.rs
//! Sync-related Tauri commands
//!
//! Provides IPC commands for cloud sync operations.

use crate::sync::{ConflictResolution, SyncConfig, SyncConflict, SyncManagerState, SyncState};
use serde::{Deserialize, Serialize};
use tauri::State;

/// Input for enabling sync
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EnableSyncInput {
    /// Authentication token
    pub auth_token: String,
    /// Optional custom server URL
    pub server_url: Option<String>,
}

/// Get current sync status
#[tauri::command]
pub async fn sync_status(
    sync_manager: State<'_, SyncManagerState>,
) -> Result<SyncState, String> {
    Ok(sync_manager.get_state().await)
}

/// Enable sync with authentication
#[tauri::command]
pub async fn sync_enable(
    input: EnableSyncInput,
    sync_manager: State<'_, SyncManagerState>,
) -> Result<(), String> {
    // Update config if custom server URL provided
    if let Some(server_url) = input.server_url {
        let mut config = sync_manager.get_config().await;
        config.server_url = server_url;
        sync_manager.set_config(config).await;
    }

    sync_manager.enable(input.auth_token).await
}

/// Disable sync
#[tauri::command]
pub async fn sync_disable(sync_manager: State<'_, SyncManagerState>) -> Result<(), String> {
    sync_manager.disable().await;
    Ok(())
}

/// Trigger manual sync
#[tauri::command]
pub async fn sync_now(sync_manager: State<'_, SyncManagerState>) -> Result<SyncResult, String> {
    let state = sync_manager.get_state().await;

    if !state.enabled {
        return Err("Sync is not enabled".to_string());
    }

    // For now, return a placeholder result
    // In a full implementation, this would call push_changes and pull_changes
    let pending = sync_manager.pending_count().await?;

    Ok(SyncResult {
        pushed: pending,
        pulled: 0,
        conflicts: 0,
    })
}

/// Result of a sync operation
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SyncResult {
    /// Number of changes pushed
    pub pushed: usize,
    /// Number of changes pulled
    pub pulled: usize,
    /// Number of conflicts detected
    pub conflicts: usize,
}

/// Get all unresolved conflicts
#[tauri::command]
pub async fn sync_get_conflicts(
    sync_manager: State<'_, SyncManagerState>,
) -> Result<Vec<SyncConflict>, String> {
    sync_manager.get_conflicts().await
}

/// Resolve a sync conflict
#[tauri::command]
pub async fn sync_resolve_conflict(
    conflict_id: String,
    resolution: ConflictResolution,
    sync_manager: State<'_, SyncManagerState>,
) -> Result<(), String> {
    sync_manager.resolve_conflict(&conflict_id, resolution).await?;
    Ok(())
}

/// Get sync configuration
#[tauri::command]
pub async fn sync_get_config(
    sync_manager: State<'_, SyncManagerState>,
) -> Result<SyncConfig, String> {
    Ok(sync_manager.get_config().await)
}

/// Update sync configuration
#[tauri::command]
pub async fn sync_set_config(
    config: SyncConfig,
    sync_manager: State<'_, SyncManagerState>,
) -> Result<(), String> {
    sync_manager.set_config(config).await;
    Ok(())
}

/// Get pending change count
#[tauri::command]
pub async fn sync_pending_count(
    sync_manager: State<'_, SyncManagerState>,
) -> Result<usize, String> {
    sync_manager.pending_count().await
}

File: apps/desktop/src/lib/api/sync.ts
/**
 * Sync API - Tauri IPC wrappers for sync operations
 */

import { invoke } from "@tauri-apps/api/core"

/**
 * Sync state
 */
export interface SyncState {
  /** Whether sync is enabled */
  enabled: boolean
  /** Whether currently syncing */
  syncing: boolean
  /** Last successful sync timestamp */
  last_sync: number | null
  /** Number of pending local changes */
  pending_changes: number
  /** Number of unresolved conflicts */
  conflicts_count: number
  /** Last error message if any */
  last_error: string | null
  /** Remote sync cursor */
  sync_cursor: string | null
}

/**
 * Sync configuration
 */
export interface SyncConfig {
  /** Sync server URL */
  server_url: string
  /** Authentication token */
  auth_token: string | null
  /** Whether to auto-sync on changes */
  auto_sync: boolean
  /** Auto-sync interval in seconds (0 = disabled) */
  sync_interval: number
  /** Unique device identifier */
  device_id: string
}

/**
 * Sync conflict
 */
export interface SyncConflict {
  /** Unique conflict ID */
  id: string
  /** Type of entity in conflict */
  entity_type: "memory" | "space" | "tag"
  /** ID of the conflicting entity */
  entity_id: string
  /** Local version of the entity (serialized) */
  local_data: string
  /** Remote version of the entity (serialized) */
  remote_data: string
  /** When the local change occurred */
  local_timestamp: number
  /** When the remote change occurred */
  remote_timestamp: number
  /** When the conflict was detected */
  detected_at: number
}

/**
 * Conflict resolution strategy
 */
export type ConflictResolution = "keep_local" | "keep_remote" | "keep_both"

/**
 * Result of a sync operation
 */
export interface SyncResult {
  /** Number of changes pushed */
  pushed: number
  /** Number of changes pulled */
  pulled: number
  /** Number of conflicts detected */
  conflicts: number
}

/**
 * Get current sync status
 */
export const syncStatus = (): Promise<SyncState> => invoke("sync_status")

/**
 * Enable sync with authentication
 */
export const syncEnable = (
  authToken: string,
  serverUrl?: string
): Promise<void> =>
  invoke("sync_enable", {
    input: { auth_token: authToken, server_url: serverUrl },
  })

/**
 * Disable sync
 */
export const syncDisable = (): Promise<void> => invoke("sync_disable")

/**
 * Trigger manual sync
 */
export const syncNow = (): Promise<SyncResult> => invoke("sync_now")

/**
 * Get all unresolved conflicts
 */
export const syncGetConflicts = (): Promise<SyncConflict[]> =>
  invoke("sync_get_conflicts")

/**
 * Resolve a sync conflict
 */
export const syncResolveConflict = (
  conflictId: string,
  resolution: ConflictResolution
): Promise<void> => invoke("sync_resolve_conflict", { conflictId, resolution })

/**
 * Get sync configuration
 */
export const syncGetConfig = (): Promise<SyncConfig> =>
  invoke("sync_get_config")

/**
 * Update sync configuration
 */
export const syncSetConfig = (config: SyncConfig): Promise<void> =>
  invoke("sync_set_config", { config })

/**
 * Get pending change count
 */
export const syncPendingCount = (): Promise<number> =>
  invoke("sync_pending_count")
